sys/include/net/gcoap.h

```diff
@@ -0,0 +1,419 @@
+#define GCOAP_MSG_TYPE_TIMEOUT    (0x1501)
+
+/**
+ * @brief Identifies a request to interrupt listening for an incoming message
+ *        on a sock.
+ *
+ * Allows the event loop to process IPC messages.
+ */
+#define GCOAP_MSG_TYPE_INTR    (0x1502)
+
+/**
+ * @brief  A modular collection of resources for a server
+ */
+typedef struct gcoap_listener {
+    coap_resource_t *resources;   /**< First element in the array of resources;
+                                       must order alphabetically */
+    size_t resources_len;         /**< Length of array */
+    struct gcoap_listener *next;  /**< Next listener in list */
+} gcoap_listener_t;
+
+/**
+ * @brief  Handler function for a server response, including the state for the
+ *         originating request.
+ *
+ * If request timed out, the packet header is for the request.
+ */
+typedef void (*gcoap_resp_handler_t)(unsigned req_state, coap_pkt_t* pdu);
+
+/**
+ * @brief  Memo to handle a response for a request
+ */
+typedef struct {
+    unsigned state;                     /**< State of this memo, a GCOAP_MEMO... */
+    uint8_t hdr_buf[GCOAP_HEADER_MAXLEN];
+                                        /**< Stores a copy of the request header */
+    gcoap_resp_handler_t resp_handler;  /**< Callback for the response */
+    xtimer_t response_timer;            /**< Limits wait for response */
+    msg_t timeout_msg;                  /**< For response timer */
+} gcoap_request_memo_t;
+
+/**
+ * @brief  Container for the state of gcoap itself
+ */
+typedef struct {
+    gcoap_listener_t *listeners;       /**< List of registered listeners */
+    gcoap_request_memo_t open_reqs[GCOAP_REQ_WAITING_MAX];
+                                       /**< Storage for open requests; if first
+                                            byte of an entry is zero, the entry
+                                            is available */
+    uint16_t last_message_id;          /**< Last message ID used */
+} gcoap_state_t;
+
+/**
+ * @brief   Initializes the gcoap thread and device.
+ *
+ * Must call once before first use.
+ *
+ * @return  PID of the gcoap thread on success.
+ * @return  -EEXIST, if thread already has been created.
+ * @return  -EINVAL, if the IP port already is in use.
+ */
+kernel_pid_t gcoap_init(void);
+
+/**
+ * @brief   Starts listening for resource paths.
+ *
+ * @param listener Listener containing the resources.
+ */
+void gcoap_register_listener(gcoap_listener_t *listener);
+
+/**
+ * @brief  Initializes a CoAP request PDU on a buffer.
+ *
+ * @param[in] pdu Request metadata
+ * @param[in] buf Buffer containing the PDU
+ * @param[in] len Length of the buffer
+ * @param[in] code Request code
+ * @param[in] path Resource path
+ *
+ * @return 0 on success
+ * @return < 0 on error
+ */
+int gcoap_req_init(coap_pkt_t *pdu, uint8_t *buf, size_t len, unsigned code,
+                                                              char *path);
```

Comment:
The main goal is readability. If the declaration extends beyond @80 chars, I right-align the following parameters below the last one on the first line. I think this block-like look is easier to read than wrapping the following arguments starting under the first argument on the first line.